---
title: "Estimating ancestry proportions"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Estimating ancestry proportions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

As genome-wide association studies (GWAS) continue to expand in scope and diversity, the precise quantification of a study's ancestral makeup has become increasingly important. The `afmix()` function of the GAUSS package, provides users with a robust method for accurately estimating these ancestry proportions within multi-ethnic GWAS, utilizing solely summary statistics data. This vignette offers a step-by-step guide on how to effectively employ the `afmix()` function.

## Load necessary packages
```{r setup, warning=FALSE, message=FALSE}
library(gauss)
library(dplyr)
library(data.table)
library(kableExtra)
```

## Overview of afmix()

The `afmix()` function requires four specific arguments:

* `input_file`: File name of the allele frequency data
* `reference_index_file`: File name of reference panel index data
* `reference_data_file`: File name of reference panel data
* `reference_pop_desc_file`: File name of reference panel population description data


Upon execution, `afmix()` function return a data.frame comprising three columns:

* `sup.pop`: super population name
* `pop`: population name
* `wgt`: population proportion


## Preparing the Input Data

### Allele Frequency Data

The allele frequency data file should be formatted as a space-delimited text file,  containing six specific columns: `rsid` (SNP ID), `chr` (chromosome number), `bp` (base pair position), `a1` (reference allele), `a2` (alternative allele), and `af1` (reference allele frequency). In this vignette, we will be using allele frequency data of Chromosome 22 from the Psychiatric Genomic Consortium's Phase 2 Schizophrenia (PGC SCZ2) GWAS.

Below, we specify the path to the allele frequency data file.
```{r prep_input}
# Path to the input file
input_file <- "../data/PGC2_Chr22_ilmn1M_AF1.txt"

# Input file should include six columns (rsid, chr, bp, a1, a2, and af1)
input.data <- fread(input_file, header = TRUE)
head(input.data)
```

### Reference Panel Data Files

Next, we assign the paths to reference panel data files. In this example, we use [the 33KG reference panel](https://statsleelab.github.io/gauss/articles/ref_33KG.html).
```{r prep_ref_path}
# Paths to the reference files (replace these with your actual paths)
reference_index_file <-"../ref/Human/33KG/33kg_index.gz"
reference_data_file <- "../ref/Human/33KG/33kg_geno.gz"
reference_pop_desc_file<-"../ref/Human/33KG/33kg_pop_desc.txt"
```


## Running afmix()

With the necessary arguments and data files in place, we are ready to run the `afmix()` function to compute ancestry proportion of `PGC SCZ2 GWAS`.
```{r run_function, message=FALSE, warning=FALSE, results='hide'}
wgt.df <- afmix(input_file=input_file,
                reference_index_file = reference_index_file,
                reference_data_file = reference_data_file,
                reference_pop_desc_file = reference_pop_desc_file)
```

## Results: Estimated Ancestry Proportions

Here, we display the estimated ancestry proportions in a HTML table.

```{r}
wgt.df %>% kable("html")
```

## Summarizing Ancestry Proportions by Super Population

Here, we calculate the total proportion for each super population and present the results in a table.
```{r sup_pop_prop}
wgt.df %>%
  group_by(sup.pop) %>%
  summarise(wgt=sum(wgt), .groups="drop") %>%
  kable("html")
```

```{r, message=FALSE, warning=FALSE, results=FALSE, echo=FALSE}
#head(wgt.df)
#PGC2_SCZ_ANC_Prop <- wgt.df[,-1]
#save(PGC2_SCZ_ANC_Prop, file = "data/pgc2_scz_anc_prop.RData")
```


## References

* Lee et al. DISTMIX: direct imputation of summary statistics for unmeasured SNPs from mixed ethnicity cohorts. Bioinformatics. https://doi.org/10.1093/bioinformatics/btv348
