---
title: "Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Introduction

This vignette provides a detailed guide on how to use the distmix() function to impute the association Z-scores of unmeasured SNPs in multi-ethnic cohorts. We'll cover the essential arguments needed and showcase a practical example.

## Load necessary packages
```{r setup, warning=FALSE, message=FALSE}
# Load necessary packages
library(gauss)
library(tidyverse)
library(data.table)
library(kableExtra)
```

## Function Arguments

The distmix() function requires following arguments:

1. `chr`: Chromosome number.
2. `start_bp`: Starting base pair position of the prediction window.
3. `end_bp`: Ending base pair position of the prediction window.
4. `wing_size`: Size of the area flanking the left and right of the prediction window.
5. `pop_wgt_df`: R data frame containing population IDs and wights
6. `input_file`: File name of the association Z-score data
7. `reference_index_file`: File name of reference panel index data
8. `reference_data_file`: File name of reference panel data
9. `reference_pop_desc_file`: File name of reference panel population description data
10. `af1_cutff`: Cutoff of reference allele (a1) frequency

## Preparing the Data

### Ancestry Proportion Data

The ancestry proportion data `pop_wgt_df` should include two columns with the following names: pop (population abbreviation) wgt (weight or proportion). This infomormation should be estimated using the `afmix()` function as described [here](https://statsleelab.github.io/gauss/articles/afmix_example.html). If this information is already available, load the data. 

Let's load the ancestry proportion data generated from the [vignette](https://statsleelab.github.io/gauss/articles/afmix_example.html) for `afmix()` function:

```{r prep_mix_prop}
# Load the ancestry proportion data
data("PGC2_SCZ_ANC_Prop") # data frame name: PGC2_SCZ_ANC_Prop
head(PGC2_SCZ_ANC_Prop)
```

### Association Z-score Data

The association Z-score data file should be a space-delimited text file including six columns with the following names: `rsid` (SNP ID), `chr` (chromosome number), `bp` (base pair position), `a1` (reference allele), `a2` (alternative allele), and `z` (association Z-score). 


Here, we will assign the path to the association Z-score data file:
```{r prep_input}
# Define the path to the input file
input_file <- "/Users/leed13/Desktop/GAUSS/GWAS_data/PGC/PGC2/PGC2_final/ilmn1M/pgc2.chr22.ilmn1M.Zonly.txt"

# Input file should include six columns (rsid, chr, bp, a1, a2, and z)
input_data <- fread(input_file, header = TRUE)
head(input_data)
```

### Reference Panel Data Files

Next, assign the paths to reference panel data files:
```{r prep_ref_path, eval=FALSE}
# Paths to the reference files (replace these with your actual paths)
reference_index_file<-"/path/to/33kg_index.gz"
reference_data_file <- "/path/to/33kg_geno.gz"
reference_pop_desc_file<-"path/to/33kg_pop_desc.txt"
```

```{r prep_ref, echo=FALSE}
reference_index_file<-"/Users/leed13/Desktop/GAUSS/ref/Human/33KG/chr_data/33kg_chr22_index.gz"
reference_data_file <- "/Users/leed13/Desktop/GAUSS/ref/Human/33KG/chr_data/33kg_chr22_geno.gz"
reference_pop_desc_file<-"/Users/leed13/Desktop/GAUSS/ref/Human/33KG/33kg_pop_desc.txt"
```


## Running distmix()

We will now execute the distmix() function with the necessary arguments and store the results:
```{r run_function, message=FALSE, warning=FALSE, results='hide'}
af1_cutoff = 0.001

res <- distmix(chr=22, start_bp = 26000001, end_bp = 26100000, wing_size = 100000,
              pop_wgt_df = PGC2_SCZ_ANC_Prop,
              input_file = input_file, 
              reference_index_file = reference_index_file,
              reference_data_file = reference_data_file, 
              reference_pop_desc_file = reference_pop_desc_file,
              af1_cutoff = af1_cutoff)
```

## Results

Let's display the results in a neat table:
```{r}
head(res) %>% kable("html")
```


## Manhattan Plot

Here, we create a Manhattan Plot to visually represent our results:
```{r man_plot, fig.height=6, fig.width=8}
res.info <- res %>% filter(info>0.6)
gwas.sig <- 5*10^-8
  
ggplot(res.info, aes(x = bp, y = -log10(pval), color = info)) +
  geom_point(alpha = 0.8) +
  geom_hline(aes(yintercept = -log10(gwas.sig)), 
             linetype = "dashed", 
             color = "red", 
             linewidth = 1.0) +
  scale_color_gradient(low = "blue", high = "red") +
  labs(title = "Manhattan Plot",
       x = "Base Pair Position",
       y = "-Log10(p-value)",
       color = "Info") +
  theme_minimal()
```