---
title: "Winner's curse adjustment"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Winner's curse adjustment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

In many genetic association studies that involve extensive multiple testing, it's often observed that association signals below the statistical significance threshold have a greater impact on trait variation than those that are statistically significant. Accurately quantifying these sub-threshold effects is crucial but challenging due to biases known as the "Winner's Curse." In previous work, we developed FIQT (FDR Inverse Quantile Transformation), a straightforward technique to correct for these biases. Initially, FIQT uses the False Discovery Rate (FDR) to correct the p-values of SNP associations for multiple testing. It then derives the Z-score noncentrality estimates based on the Gaussian quantiles that correspond to these adjusted p-values, while maintaining the correct sign. The FIQT approach has been conveniently incorporated into the GAUSS package as the `fiqt()` function. By simply entering a Z-score vector along with the minimum non-zero p-value generated by the qnorm() function (default value is 10^-320), this function will produce a Z-score vector that has been adjusted to account for the winner's curse biases.


## Load necessary packages
```{r setup}
library(gauss)
library(tidyverse)
library(data.table)
library(kableExtra)
```

## Example

In this example, we'll correct for the "Winner's Curse" bias using the fiqt() function applied to summary statistics from the Psychiatric Genomics Consortium Schizophrenia Phase 3 study. For demonstration purposes, we have limited our analysis to SNPs from the Illumina 1M Chip only.

```{r}
input_file <- "../data/PGC3_SCZ_ilmn1M_Z.txt"
pgc3 <- fread(input_file)
head(pgc3)
```

### Running fiqt()

```{r}
pgc3$z.wca <- fiqt(pgc3$z)
```

### Results

```{r}
head(pgc3) %>% kable("html")
```

### Original Z-scores vs FIQT Z-score noncentrality estimates

```{r, fig.height=5, fig.width=5}
ggplot() +
  geom_point(aes(x=z, y=z.wca), data=pgc3) +
  geom_abline(intercept = 0, slope = 1, color="red") +
  labs(x="Original Z-score",
      y="Estimated Z-score noncentrality",
      caption="Data: PGC SCZ3 GWAS") +
  theme_classic()
```

## Reference

* Bigdeli et al. A simple yet accurate correction for winner's curse can predict signals discovered in much larger genome scans. Bioinformatics. 2016 Sep 1;32(17):2598-603. doi: 10.1093/bioinformatics/btw303.


