<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gauss">
<title>Estimating ancestry proportions • gauss</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimating ancestry proportions">
<meta property="og:description" content="gauss">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gauss</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Get Started</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/afmix_example.html">Estimating ancestry proportions</a>
    <a class="dropdown-item" href="../articles/computeLD_example.html">Estimating linkage disequilibrium</a>
    <a class="dropdown-item" href="../articles/dist_example.html">Imputing Z-scores of unmeasured SNPs</a>
    <a class="dropdown-item" href="../articles/jepeg_example.html">Transcriptome wide association study</a>
    <a class="dropdown-item" href="../articles/gsl_installation.html">GSL Installation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-reference-panels">Reference Panels</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-reference-panels">
    <a class="dropdown-item" href="../articles/ref_33KG.html">33KG Panel</a>
    <a class="dropdown-item" href="../articles/ref_1KG.html">1KG Panel</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../#">
    <span class="fa fa-search"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/statsleelab/gauss">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Estimating ancestry proportions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/statsleelab/gauss/blob/HEAD/vignettes/afmix_example.Rmd"><code>vignettes/afmix_example.Rmd</code></a></small>
      <div class="d-none name"><code>afmix_example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>As genome-wide association studies (GWAS) continue to expand in scope
and diversity, the precise quantification of a study’s ancestral makeup
has become increasingly important. The <code><a href="../reference/afmix.html">afmix()</a></code> function of
the GAUSS package, provides users with a robust method for accurately
estimating these ancestry proportions within multi-ethnic GWAS,
utilizing solely summary statistics data. This vignette offers a
step-by-step guide on how to effectively employ the <code><a href="../reference/afmix.html">afmix()</a></code>
function.</p>
</div>
<div class="section level2">
<h2 id="load-necessary-packages">Load necessary packages<a class="anchor" aria-label="anchor" href="#load-necessary-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statsleelab/gauss">gauss</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="overview-of-afmix">Overview of afmix()<a class="anchor" aria-label="anchor" href="#overview-of-afmix"></a>
</h2>
<p>The <code><a href="../reference/afmix.html">afmix()</a></code> function requires four specific
arguments:</p>
<ul>
<li>
<code>input_file</code>: File name of the allele frequency data</li>
<li>
<code>reference_index_file</code>: File name of reference panel
index data</li>
<li>
<code>reference_data_file</code>: File name of reference panel
data</li>
<li>
<code>reference_pop_desc_file</code>: File name of reference panel
population description data</li>
</ul>
<p>Upon execution, <code><a href="../reference/afmix.html">afmix()</a></code> function return a data.frame
comprising three columns:</p>
<ul>
<li>
<code>sup.pop</code>: super population name</li>
<li>
<code>pop</code>: population name</li>
<li>
<code>wgt</code>: population proportion</li>
</ul>
</div>
<div class="section level2">
<h2 id="preparing-the-input-data">Preparing the Input Data<a class="anchor" aria-label="anchor" href="#preparing-the-input-data"></a>
</h2>
<div class="section level3">
<h3 id="allele-frequency-data">Allele Frequency Data<a class="anchor" aria-label="anchor" href="#allele-frequency-data"></a>
</h3>
<p>The allele frequency data file should be formatted as a
space-delimited text file, containing six specific columns:
<code>rsid</code> (SNP ID), <code>chr</code> (chromosome number),
<code>bp</code> (base pair position), <code>a1</code> (reference
allele), <code>a2</code> (alternative allele), and <code>af1</code>
(reference allele frequency). In this vignette, we will be using allele
frequency data of Chromosome 22 from the Psychiatric Genomic
Consortium’s Phase 2 Schizophrenia (PGC SCZ2) GWAS.</p>
<p>Below, we specify the path to the allele frequency data file.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to the input file</span></span>
<span><span class="va">input_file</span> <span class="op">&lt;-</span> <span class="st">"../data/PGC2_Chr22_ilmn1M_AF1.txt"</span></span>
<span></span>
<span><span class="co"># Input file should include six columns (rsid, chr, bp, a1, a2, and af1)</span></span>
<span><span class="va">input.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">input_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         rsid chr       bp a1 a2       af1</span></span>
<span><span class="co">#&gt; 1: rs1000427  22 36890105  A  G 0.1159800</span></span>
<span><span class="co">#&gt; 2: rs1000470  22 24026845  A  C 0.1240050</span></span>
<span><span class="co">#&gt; 3: rs1000539  22 20202729  A  G 0.2660650</span></span>
<span><span class="co">#&gt; 4:   rs10009  22 22051709  A  G 0.6319500</span></span>
<span><span class="co">#&gt; 5: rs1001022  22 26403488  T  C 0.9570250</span></span>
<span><span class="co">#&gt; 6: rs1001213  22 34131736  A  G 0.0667805</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reference-panel-data-files">Reference Panel Data Files<a class="anchor" aria-label="anchor" href="#reference-panel-data-files"></a>
</h3>
<p>Next, we assign the paths to reference panel data files. In this
example, we use <a href="https://statsleelab.github.io/gauss/articles/ref_33KG.html" class="external-link">the
33KG reference panel</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Paths to the reference files (replace these with your actual paths)</span></span>
<span><span class="va">reference_index_file</span> <span class="op">&lt;-</span><span class="st">"../ref/Human/33KG/33kg_index.gz"</span></span>
<span><span class="va">reference_data_file</span> <span class="op">&lt;-</span> <span class="st">"../ref/Human/33KG/33kg_geno.gz"</span></span>
<span><span class="va">reference_pop_desc_file</span><span class="op">&lt;-</span><span class="st">"../ref/Human/33KG/33kg_pop_desc.txt"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-afmix">Running afmix()<a class="anchor" aria-label="anchor" href="#running-afmix"></a>
</h2>
<p>With the necessary arguments and data files in place, we are ready to
run the <code><a href="../reference/afmix.html">afmix()</a></code> function to compute ancestry proportion of
<code>PGC SCZ2 GWAS</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgt.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/afmix.html">afmix</a></span><span class="op">(</span>input_file<span class="op">=</span><span class="va">input_file</span>,</span>
<span>                reference_index_file <span class="op">=</span> <span class="va">reference_index_file</span>,</span>
<span>                reference_data_file <span class="op">=</span> <span class="va">reference_data_file</span>,</span>
<span>                reference_pop_desc_file <span class="op">=</span> <span class="va">reference_pop_desc_file</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="results-estimated-ancestry-proportions">Results: Estimated Ancestry Proportions<a class="anchor" aria-label="anchor" href="#results-estimated-ancestry-proportions"></a>
</h2>
<p>Here, we display the estimated ancestry proportions in a HTML
table.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgt.df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
sup.pop
</th>
<th style="text-align:left;">
pop
</th>
<th style="text-align:right;">
wgt
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
AFR
</td>
<td style="text-align:left;">
ACB
</td>
<td style="text-align:right;">
0.006
</td>
</tr>
<tr>
<td style="text-align:left;">
AFR
</td>
<td style="text-align:left;">
ASW
</td>
<td style="text-align:right;">
0.036
</td>
</tr>
<tr>
<td style="text-align:left;">
SAS
</td>
<td style="text-align:left;">
BEB
</td>
<td style="text-align:right;">
0.005
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
CCE
</td>
<td style="text-align:right;">
0.008
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
CCS
</td>
<td style="text-align:right;">
0.004
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
CDX
</td>
<td style="text-align:right;">
0.018
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
CEU
</td>
<td style="text-align:right;">
0.165
</td>
</tr>
<tr>
<td style="text-align:left;">
AMR
</td>
<td style="text-align:left;">
CLM
</td>
<td style="text-align:right;">
0.025
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
CNE
</td>
<td style="text-align:right;">
0.003
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
CSE
</td>
<td style="text-align:right;">
0.012
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
FIN
</td>
<td style="text-align:right;">
0.138
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
GBR
</td>
<td style="text-align:right;">
0.165
</td>
</tr>
<tr>
<td style="text-align:left;">
SAS
</td>
<td style="text-align:left;">
GIH
</td>
<td style="text-align:right;">
0.006
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
IBS
</td>
<td style="text-align:right;">
0.099
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
JPT
</td>
<td style="text-align:right;">
0.011
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:left;">
KHV
</td>
<td style="text-align:right;">
0.017
</td>
</tr>
<tr>
<td style="text-align:left;">
AMR
</td>
<td style="text-align:left;">
MXL
</td>
<td style="text-align:right;">
0.030
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
ORK
</td>
<td style="text-align:right;">
0.166
</td>
</tr>
<tr>
<td style="text-align:left;">
SAS
</td>
<td style="text-align:left;">
PJL
</td>
<td style="text-align:right;">
0.016
</td>
</tr>
<tr>
<td style="text-align:left;">
AMR
</td>
<td style="text-align:left;">
PUR
</td>
<td style="text-align:right;">
0.045
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:left;">
TSI
</td>
<td style="text-align:right;">
0.086
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="summarizing-ancestry-proportions-by-super-population">Summarizing Ancestry Proportions by Super Population<a class="anchor" aria-label="anchor" href="#summarizing-ancestry-proportions-by-super-population"></a>
</h2>
<p>Here, we calculate the total proportion for each super population and
present the results in a table.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wgt.df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sup.pop</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>wgt<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">wgt</span><span class="op">)</span>, .groups<span class="op">=</span><span class="st">"drop"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
sup.pop
</th>
<th style="text-align:right;">
wgt
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
AFR
</td>
<td style="text-align:right;">
0.042
</td>
</tr>
<tr>
<td style="text-align:left;">
AMR
</td>
<td style="text-align:right;">
0.100
</td>
</tr>
<tr>
<td style="text-align:left;">
ASN
</td>
<td style="text-align:right;">
0.073
</td>
</tr>
<tr>
<td style="text-align:left;">
EUR
</td>
<td style="text-align:right;">
0.819
</td>
</tr>
<tr>
<td style="text-align:left;">
SAS
</td>
<td style="text-align:right;">
0.027
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Lee et al. DISTMIX: direct imputation of summary statistics for
unmeasured SNPs from mixed ethnicity cohorts. Bioinformatics. <a href="https://doi.org/10.1093/bioinformatics/btv348" class="external-link uri">https://doi.org/10.1093/bioinformatics/btv348</a>
</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/dleelab" class="external-link">Donghyung Lee</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
