<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gauss">
<title>Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts • gauss</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts">
<meta property="og:description" content="gauss">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gauss</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/afmix_example.html">Estimating ancestry proportions</a>
    <a class="dropdown-item" href="../articles/dist_example.html">Imputing Z-scores of unmeasured SNPs</a>
    <a class="dropdown-item" href="../articles/distmix_example.html">Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/statsleelab/gauss/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Imputing Z-scores of unmeasured SNPs from multi-ethnic cohorts</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/statsleelab/gauss/blob/HEAD/vignettes/distmix_example.Rmd" class="external-link"><code>vignettes/distmix_example.Rmd</code></a></small>
      <div class="d-none name"><code>distmix_example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a detailed guide on how to use the distmix()
function to impute the association Z-scores of unmeasured SNPs in
multi-ethnic cohorts. We’ll cover the essential arguments needed and
showcase a practical example.</p>
</div>
<div class="section level2">
<h2 id="load-necessary-packages">Load necessary packages<a class="anchor" aria-label="anchor" href="#load-necessary-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statsleelab/gauss" class="external-link">gauss</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="function-arguments">Function Arguments<a class="anchor" aria-label="anchor" href="#function-arguments"></a>
</h2>
<p>The distmix() function requires following arguments:</p>
<ol style="list-style-type: decimal">
<li>
<code>chr</code>: Chromosome number.</li>
<li>
<code>start_bp</code>: Starting base pair position of the prediction
window.</li>
<li>
<code>end_bp</code>: Ending base pair position of the prediction
window.</li>
<li>
<code>wing_size</code>: Size of the area flanking the left and right
of the prediction window.</li>
<li>
<code>pop_wgt_df</code>: R data frame containing population IDs and
wights</li>
<li>
<code>input_file</code>: File name of the association Z-score
data</li>
<li>
<code>reference_index_file</code>: File name of reference panel
index data</li>
<li>
<code>reference_data_file</code>: File name of reference panel
data</li>
<li>
<code>reference_pop_desc_file</code>: File name of reference panel
population description data</li>
<li>
<code>af1_cutff</code>: Cutoff of reference allele (a1)
frequency</li>
</ol>
</div>
<div class="section level2">
<h2 id="preparing-the-data">Preparing the Data<a class="anchor" aria-label="anchor" href="#preparing-the-data"></a>
</h2>
<div class="section level3">
<h3 id="ancestry-proportion-data">Ancestry Proportion Data<a class="anchor" aria-label="anchor" href="#ancestry-proportion-data"></a>
</h3>
<p>The ancestry proportion data <code>pop_wgt_df</code> should include
two columns with the following names: pop (population abbreviation) wgt
(weight or proportion). This infomormation should be estimated using the
<code><a href="../reference/afmix.html">afmix()</a></code> function as described <a href="https://statsleelab.github.io/gauss/articles/afmix_example.html" class="external-link">here</a>.
If this information is already available, load the data.</p>
<p>Let’s load the ancestry proportion data generated from the <a href="https://statsleelab.github.io/gauss/articles/afmix_example.html" class="external-link">vignette</a>
for <code><a href="../reference/afmix.html">afmix()</a></code> function:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the ancestry proportion data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"PGC2_SCZ_ANC_Prop"</span><span class="op">)</span> <span class="co"># data frame name: PGC2_SCZ_ANC_Prop</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PGC2_SCZ_ANC_Prop</span><span class="op">)</span></span>
<span><span class="co">#&gt;   pop   wgt</span></span>
<span><span class="co">#&gt; 1 ACB 0.006</span></span>
<span><span class="co">#&gt; 2 ASW 0.036</span></span>
<span><span class="co">#&gt; 3 BEB 0.005</span></span>
<span><span class="co">#&gt; 4 CCE 0.008</span></span>
<span><span class="co">#&gt; 5 CCS 0.004</span></span>
<span><span class="co">#&gt; 6 CDX 0.018</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="association-z-score-data">Association Z-score Data<a class="anchor" aria-label="anchor" href="#association-z-score-data"></a>
</h3>
<p>The association Z-score data file should be a space-delimited text
file including six columns with the following names: <code>rsid</code>
(SNP ID), <code>chr</code> (chromosome number), <code>bp</code> (base
pair position), <code>a1</code> (reference allele), <code>a2</code>
(alternative allele), and <code>z</code> (association Z-score).</p>
<p>Here, we will assign the path to the association Z-score data
file:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the path to the input file</span></span>
<span><span class="va">input_file</span> <span class="op">&lt;-</span> <span class="st">"/Users/leed13/Desktop/GAUSS/GWAS_data/PGC/PGC2/PGC2_final/ilmn1M/pgc2.chr22.ilmn1M.Zonly.txt"</span></span>
<span></span>
<span><span class="co"># Input file should include six columns (rsid, chr, bp, a1, a2, and z)</span></span>
<span><span class="va">input_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">input_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         rsid chr       bp a1 a2          z</span></span>
<span><span class="co">#&gt; 1: rs1000427  22 36890105  A  G -1.4969741</span></span>
<span><span class="co">#&gt; 2: rs1000470  22 24026845  A  C  1.8407377</span></span>
<span><span class="co">#&gt; 3: rs1000539  22 20202729  A  G  2.1683772</span></span>
<span><span class="co">#&gt; 4:   rs10009  22 22051709  A  G -0.0556759</span></span>
<span><span class="co">#&gt; 5: rs1001022  22 26403488  T  C  2.9694844</span></span>
<span><span class="co">#&gt; 6: rs1001213  22 34131736  A  G -0.8957244</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reference-panel-data-files">Reference Panel Data Files<a class="anchor" aria-label="anchor" href="#reference-panel-data-files"></a>
</h3>
<p>Next, assign the paths to reference panel data files:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Paths to the reference files (replace these with your actual paths)</span></span>
<span><span class="va">reference_index_file</span><span class="op">&lt;-</span><span class="st">"/path/to/33kg_index.gz"</span></span>
<span><span class="va">reference_data_file</span> <span class="op">&lt;-</span> <span class="st">"/path/to/33kg_geno.gz"</span></span>
<span><span class="va">reference_pop_desc_file</span><span class="op">&lt;-</span><span class="st">"path/to/33kg_pop_desc.txt"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-distmix">Running distmix()<a class="anchor" aria-label="anchor" href="#running-distmix"></a>
</h2>
<p>We will now execute the distmix() function with the necessary
arguments and store the results:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af1_cutoff</span> <span class="op">=</span> <span class="fl">0.001</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distmix.html">distmix</a></span><span class="op">(</span>chr<span class="op">=</span><span class="fl">22</span>, start_bp <span class="op">=</span> <span class="fl">26000001</span>, end_bp <span class="op">=</span> <span class="fl">26100000</span>, wing_size <span class="op">=</span> <span class="fl">100000</span>,</span>
<span>              pop_wgt_df <span class="op">=</span> <span class="va">PGC2_SCZ_ANC_Prop</span>,</span>
<span>              input_file <span class="op">=</span> <span class="va">input_file</span>, </span>
<span>              reference_index_file <span class="op">=</span> <span class="va">reference_index_file</span>,</span>
<span>              reference_data_file <span class="op">=</span> <span class="va">reference_data_file</span>, </span>
<span>              reference_pop_desc_file <span class="op">=</span> <span class="va">reference_pop_desc_file</span>,</span>
<span>              af1_cutoff <span class="op">=</span> <span class="va">af1_cutoff</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>Let’s display the results in a neat table:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
rsid
</th>
<th style="text-align:right;">
chr
</th>
<th style="text-align:right;">
bp
</th>
<th style="text-align:left;">
a1
</th>
<th style="text-align:left;">
a2
</th>
<th style="text-align:right;">
af1mix
</th>
<th style="text-align:right;">
z
</th>
<th style="text-align:right;">
pval
</th>
<th style="text-align:right;">
info
</th>
<th style="text-align:right;">
type
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
rs7284433
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000039
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.0042145
</td>
<td style="text-align:right;">
-0.1186885
</td>
<td style="text-align:right;">
0.9055222
</td>
<td style="text-align:right;">
0.5621959
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
rs144526288
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000077
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
0.0034800
</td>
<td style="text-align:right;">
-0.8356092
</td>
<td style="text-align:right;">
0.4033748
</td>
<td style="text-align:right;">
0.2646915
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
rs77818143
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000215
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.0018339
</td>
<td style="text-align:right;">
0.5727715
</td>
<td style="text-align:right;">
0.5667994
</td>
<td style="text-align:right;">
0.0330793
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
rs146646304
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000267
</td>
<td style="text-align:left;">
T
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:right;">
0.0020257
</td>
<td style="text-align:right;">
-0.3280241
</td>
<td style="text-align:right;">
0.7428934
</td>
<td style="text-align:right;">
0.3598793
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
rs542157098
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000579
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
0.0055346
</td>
<td style="text-align:right;">
-1.5095628
</td>
<td style="text-align:right;">
0.1311550
</td>
<td style="text-align:right;">
0.4445158
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
rs16980509
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
26000651
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:right;">
0.0712967
</td>
<td style="text-align:right;">
-1.0535260
</td>
<td style="text-align:right;">
0.2921000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="manhattan-plot">Manhattan Plot<a class="anchor" aria-label="anchor" href="#manhattan-plot"></a>
</h2>
<p>Here, we create a Manhattan Plot to visually represent our
results:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.info</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">info</span><span class="op">&gt;</span><span class="fl">0.6</span><span class="op">)</span></span>
<span><span class="va">gwas.sig</span> <span class="op">&lt;-</span> <span class="fl">5</span><span class="op">*</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">8</span></span>
<span>  </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res.info</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bp</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">info</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">gwas.sig</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             linetype <span class="op">=</span> <span class="st">"dashed"</span>, </span>
<span>             color <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>             linewidth <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Manhattan Plot"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"-Log10(p-value)"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"Info"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="distmix_example_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Donghyung Lee.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
