<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gauss">
<title>Transcriptome wide association study with GAUSS • gauss</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Transcriptome wide association study with GAUSS">
<meta property="og:description" content="gauss">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gauss</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">Get Started</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/afmix_example.html">Estimating ancestry proportions</a>
    <a class="dropdown-item" href="../articles/computeLD_example.html">Estimating linkage disequilibrium</a>
    <a class="dropdown-item" href="../articles/dist_example.html">Imputing Z-scores of unmeasured SNPs</a>
    <a class="dropdown-item" href="../articles/jepeg_example.html">Transcriptome wide association study</a>
    <a class="dropdown-item" href="../articles/gsl_installation.html">GSL Installation</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-reference-panels">Reference Panels</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-reference-panels">
    <a class="dropdown-item" href="../articles/ref_33KG.html">33KG Panel</a>
    <a class="dropdown-item" href="../articles/ref_1KG.html">1KG Panel</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../#">
    <span class="fa fa-search"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/statsleelab/gauss">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Transcriptome wide association study with GAUSS</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/statsleelab/gauss/blob/HEAD/vignettes/jepeg_example.Rmd"><code>vignettes/jepeg_example.Rmd</code></a></small>
      <div class="d-none name"><code>jepeg_example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Transcriptome-Wide Association Studies (TWAS) serve as a powerful
framework for identifying the functional relationships between genetic
variations and complex traits. By integrating results from Genome-Wide
Association Studies (GWAS) with rich functional annotations, TWAS
provides a more comprehensive view of the genetic architecture of traits
and diseases. GAUSS offers a robust suite of tools to streamline and
enhance TWAS analyses. Two of its specialized functions,
<code><a href="../reference/jepeg.html">jepeg()</a></code> and <code><a href="../reference/jepegmix.html">jepegmix()</a></code>, offer tailored
solutions for homogeneous and multi-ethnic cohorts, respectively. This
vignette aims to be your comprehensive guide, walking you through the
hands-on application of <code><a href="../reference/jepeg.html">jepeg()</a></code> and <code><a href="../reference/jepegmix.html">jepegmix()</a></code>
in conducting TWAS.</p>
</div>
<div class="section level2">
<h2 id="load-necessary-packages">Load necessary packages<a class="anchor" aria-label="anchor" href="#load-necessary-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statsleelab/gauss">gauss</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preparing-the-input-data">Preparing the Input Data<a class="anchor" aria-label="anchor" href="#preparing-the-input-data"></a>
</h2>
<p>Both <code><a href="../reference/jepeg.html">jepeg()</a></code> and <code><a href="../reference/jepegmix.html">jepegmix()</a></code> functions
necessitate three core input datasets:</p>
<ul>
<li>Association Z-score Data</li>
<li>SNP Annotation Data</li>
<li>Reference Panel Data</li>
</ul>
<div class="section level3">
<h3 id="association-z-score-data">Association Z-score Data<a class="anchor" aria-label="anchor" href="#association-z-score-data"></a>
</h3>
<p>This data should be in the form of a space-delimited text file
including six columns with the following names:</p>
<ul>
<li>
<code>rsid</code>: SNP ID</li>
<li>
<code>chr</code>: Chromosome number</li>
<li>
<code>bp</code>: Base pair position</li>
<li>
<code>a1</code>: Reference allele</li>
<li>
<code>a2</code>: Alternative allele</li>
<li>
<code>z</code>: Association Z-score</li>
</ul>
<p>For this example, we will use the Psychiatric Genomic Consortium’s
Phase 2 Schizophrenia (PGC SCZ2) GWAS dataset.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to the input file</span></span>
<span><span class="va">input_file</span> <span class="op">&lt;-</span> <span class="st">"../data/PGC2_Chr22_ilmn1M_Z.txt"</span></span>
<span></span>
<span><span class="co"># Input file should include six columns (rsid, chr, bp, a1, a2, and z)</span></span>
<span><span class="va">input.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">input_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         rsid chr       bp a1 a2          z</span></span>
<span><span class="co">#&gt; 1: rs1000427  22 36890105  A  G -1.4969741</span></span>
<span><span class="co">#&gt; 2: rs1000470  22 24026845  A  C  1.8407377</span></span>
<span><span class="co">#&gt; 3: rs1000539  22 20202729  A  G  2.1683772</span></span>
<span><span class="co">#&gt; 4:   rs10009  22 22051709  A  G -0.0556759</span></span>
<span><span class="co">#&gt; 5: rs1001022  22 26403488  T  C  2.9694844</span></span>
<span><span class="co">#&gt; 6: rs1001213  22 34131736  A  G -0.8957244</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="snp-annotation-data">SNP Annotation Data<a class="anchor" aria-label="anchor" href="#snp-annotation-data"></a>
</h3>
<p>The SNP Annotation Data should be placed in an accessible directory,
as shown below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Paths to the annotation file (replace these with your actual paths)</span></span>
<span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="st">"../data/JEPEG_SNP_Annotation.v1.0.txt"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reference-panel-data">Reference Panel Data<a class="anchor" aria-label="anchor" href="#reference-panel-data"></a>
</h3>
<p>We will be using <a href="https://statsleelab.github.io/gauss/articles/ref_33KG.html" class="external-link">the
33KG reference panel</a>. Make sure to replace the file paths below with
those of your actual reference panel files.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reference_index_file</span> <span class="op">&lt;-</span><span class="st">"../ref/Human/33KG/33kg_index.gz"</span></span>
<span><span class="va">reference_data_file</span> <span class="op">&lt;-</span> <span class="st">"../ref/Human/33KG/33kg_geno.gz"</span></span>
<span><span class="va">reference_pop_desc_file</span><span class="op">&lt;-</span><span class="st">"../ref/Human/33KG/33kg_pop_desc.txt"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-jepeg-function">The <code>jepeg()</code> function<a class="anchor" aria-label="anchor" href="#the-jepeg-function"></a>
</h2>
<p>The <code><a href="../reference/jepeg.html">jepeg()</a></code> function enables researchers to perform TWAS
by assessing the joint effect of expression quantitative trait loci
(eQTLs) and other functional variants within a specific gene on
phenotype. It requires an SNP annotation file that provides a linkage
between SNPs and their corresponding gene-level functional
characteristics. By integrating this data, jepeg() formulates gene-level
test statistics, aiming to pinpoint genes whose expression levels might
be modulated by genetic variations.”</p>
<div class="section level3">
<h3 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a>
</h3>
<p>The <code><a href="../reference/jepeg.html">jepeg()</a></code> function requires the following
arguments:</p>
<ul>
<li>
<code>study_pop</code>: Study population group</li>
<li>
<code>input_file</code>: File name of the association Z-score
data</li>
<li>
<code>annotation</code>: File name of the SNP annotation data
set</li>
<li>
<code>reference_index_file</code>: File name of reference panel
index data</li>
<li>
<code>reference_data_file</code>: File name of reference panel
data</li>
<li>
<code>reference_pop_desc_file</code>: File name of reference panel
population description data</li>
<li>
<code>af1_cutoff</code>: Cutoff of reference allele (a1)
frequency</li>
</ul>
</div>
<div class="section level3">
<h3 id="outputs">Outputs<a class="anchor" aria-label="anchor" href="#outputs"></a>
</h3>
<p>The <code><a href="../reference/jepeg.html">jepeg()</a></code> function returns a data frame with following
columns:</p>
<ul>
<li>
<code>geneid</code>: Gene name.</li>
<li>
<code>chisq</code>: Chi-square test statistic value</li>
<li>
<code>df</code>: Degrees of freedom</li>
<li>
<code>jepeg_pval</code>: JEPEG p-value</li>
<li>
<code>num_snp</code>: Number of functional SNPs associated with
gene</li>
<li>
<code>top_categ</code>: Top functional category</li>
<li>
<code>top_categ_pval</code>: Top category p-value</li>
<li>
<code>top_snp</code>: Top functional SNP ID</li>
<li>
<code>top_snp_pval</code>: Top functional SNP p-value</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-usage">Example Usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h3>
<p>In this example, we will use the <code><a href="../reference/jepeg.html">jepeg()</a></code> function to
perform a TWAS analysis utilizing summary statistics from the PGC SCZ2
GWAS dataset. Assuming our study cohort consists of participants of
British descent, we will specify <code>study_pop = "GBR"</code>. This
will ensure the function employs genotype data for individuals from
Britain, specifically England and Scotland, as sourced from the 33KG
reference panel.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af1_cutoff</span> <span class="op">=</span> <span class="fl">0.001</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/jepeg.html">jepeg</a></span><span class="op">(</span>study_pop <span class="op">=</span> <span class="st">"GBR"</span>,</span>
<span>             input_file<span class="op">=</span><span class="va">input_file</span>,</span>
<span>             annotation_file<span class="op">=</span><span class="va">annotation_file</span>,</span>
<span>             reference_index_file <span class="op">=</span> <span class="va">reference_index_file</span>,</span>
<span>             reference_data_file <span class="op">=</span> <span class="va">reference_data_file</span>,</span>
<span>             reference_pop_desc_file <span class="op">=</span> <span class="va">reference_pop_desc_file</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">jepeg_pval</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="jepeg-results">
<code>jepeg()</code> Results<a class="anchor" aria-label="anchor" href="#jepeg-results"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
geneid
</th>
<th style="text-align:right;">
chisq
</th>
<th style="text-align:right;">
df
</th>
<th style="text-align:right;">
jepeg_pval
</th>
<th style="text-align:right;">
num_snp
</th>
<th style="text-align:left;">
top_categ
</th>
<th style="text-align:right;">
top_categ_pval
</th>
<th style="text-align:left;">
top_snp
</th>
<th style="text-align:right;">
top_snp_pval
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
85
</td>
<td style="text-align:left;">
DPYD
</td>
<td style="text-align:right;">
38.41841
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRN
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:left;">
rs3788568
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
72
</td>
<td style="text-align:left;">
CXCL14
</td>
<td style="text-align:right;">
33.98061
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRN
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:left;">
rs133047
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
93
</td>
<td style="text-align:left;">
EP300
</td>
<td style="text-align:right;">
29.29304
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.00e-07
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
1.00e-07
</td>
<td style="text-align:left;">
rs20551
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
328
</td>
<td style="text-align:left;">
WBP2NL
</td>
<td style="text-align:right;">
24.71184
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
7.00e-07
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
7.00e-07
</td>
<td style="text-align:left;">
rs2301521
</td>
<td style="text-align:right;">
1.0e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
186
</td>
<td style="text-align:left;">
NDUFA6
</td>
<td style="text-align:right;">
24.39774
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
8.00e-07
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
8.00e-07
</td>
<td style="text-align:left;">
rs1801311
</td>
<td style="text-align:right;">
2.0e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
332
</td>
<td style="text-align:left;">
ZBED4
</td>
<td style="text-align:right;">
19.38566
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.07e-05
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
1.07e-05
</td>
<td style="text-align:left;">
rs910799
</td>
<td style="text-align:right;">
3.9e-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="the-jepegmix-function">The <code>jepegmix()</code> function<a class="anchor" aria-label="anchor" href="#the-jepegmix-function"></a>
</h2>
<p>For studies that involve diverse ethnic backgrounds,
<code><a href="../reference/jepegmix.html">jepegmix()</a></code> extends the utility of <code><a href="../reference/jepeg.html">jepeg()</a></code> to
accommodate the complexities introduced by ethnic heterogeneity.</p>
<div class="section level3">
<h3 id="arguments-1">Arguments<a class="anchor" aria-label="anchor" href="#arguments-1"></a>
</h3>
<p>The <code><a href="../reference/jepegmix.html">jepegmix()</a></code> function accepts most of the same
arguments as <code><a href="../reference/jepeg.html">jepeg()</a></code>. However, it replaces the
<code>study_pop</code> argument with <code>pop_wgt_df</code>:</p>
<ul>
<li>
<code>pop_wgt_df</code>: An R data frame containing population names
and corresponding ancestry proportions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="outputs-1">Outputs<a class="anchor" aria-label="anchor" href="#outputs-1"></a>
</h3>
<p>The output of <code><a href="../reference/jepegmix.html">jepegmix()</a></code> is identical to that of
<code><a href="../reference/jepeg.html">jepeg()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="example-usage-1">Example Usage<a class="anchor" aria-label="anchor" href="#example-usage-1"></a>
</h3>
<p>Before using <code><a href="../reference/jepegmix.html">jepegmix()</a></code>, you need to prepare ancestry
proportion data. This data should be structured as a data frame
containing two columns:</p>
<ul>
<li>
<code>pop</code>: Population abbreviation</li>
<li>
<code>wgt</code>: Corresponding proportion for that population</li>
</ul>
<p>To estimate these ancestry proportions, the <code><a href="../reference/afmix.html">afmix()</a></code>
function can be employed. For a step-by-step guide on this process,
refer to the <a href="https://statsleelab.github.io/gauss/articles/afmix_example.html" class="external-link"><code>afmix()</code>
vignette</a>.</p>
<p>Here, we load pre-generated ancestry proportion data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the ancestry proportion data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"PGC2_SCZ_ANC_Prop"</span><span class="op">)</span> <span class="co"># data frame name: PGC2_SCZ_ANC_Prop</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PGC2_SCZ_ANC_Prop</span><span class="op">)</span></span>
<span><span class="co">#&gt;   pop   wgt</span></span>
<span><span class="co">#&gt; 1 ACB 0.006</span></span>
<span><span class="co">#&gt; 2 ASW 0.036</span></span>
<span><span class="co">#&gt; 3 BEB 0.005</span></span>
<span><span class="co">#&gt; 4 CCE 0.008</span></span>
<span><span class="co">#&gt; 5 CCS 0.004</span></span>
<span><span class="co">#&gt; 6 CDX 0.018</span></span></code></pre></div>
<p>Now, let’s proceed to use <code><a href="../reference/jepegmix.html">jepegmix()</a></code> to conduct TWAS.
Here, we assume the study cohort includes participants from diverse
ethnic backgrounds.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af1_cutoff</span> <span class="op">=</span> <span class="fl">0.001</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/jepegmix.html">jepegmix</a></span><span class="op">(</span>pop_wgt_df <span class="op">=</span> <span class="va">PGC2_SCZ_ANC_Prop</span>,</span>
<span>                input_file<span class="op">=</span><span class="va">input_file</span>,</span>
<span>                annotation_file<span class="op">=</span><span class="va">annotation_file</span>,</span>
<span>                reference_index_file <span class="op">=</span> <span class="va">reference_index_file</span>,</span>
<span>                reference_data_file <span class="op">=</span> <span class="va">reference_data_file</span>,</span>
<span>                reference_pop_desc_file <span class="op">=</span> <span class="va">reference_pop_desc_file</span>,</span>
<span>                af1_cutoff <span class="op">=</span> <span class="va">af1_cutoff</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">jepeg_pval</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="jepegmix-results">
<code>jepegmix()</code> Results<a class="anchor" aria-label="anchor" href="#jepegmix-results"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="st">"html"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
geneid
</th>
<th style="text-align:right;">
chisq
</th>
<th style="text-align:right;">
df
</th>
<th style="text-align:right;">
jepeg_pval
</th>
<th style="text-align:right;">
num_snp
</th>
<th style="text-align:left;">
top_categ
</th>
<th style="text-align:right;">
top_categ_pval
</th>
<th style="text-align:left;">
top_snp
</th>
<th style="text-align:right;">
top_snp_pval
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
92
</td>
<td style="text-align:left;">
DPYD
</td>
<td style="text-align:right;">
38.41841
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRN
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:left;">
rs3788568
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
75
</td>
<td style="text-align:left;">
CXCL14
</td>
<td style="text-align:right;">
33.81350
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRN
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:left;">
rs133047
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
101
</td>
<td style="text-align:left;">
EP300
</td>
<td style="text-align:right;">
29.29304
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.00e-07
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
1.00e-07
</td>
<td style="text-align:left;">
rs20551
</td>
<td style="text-align:right;">
0.0e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
361
</td>
<td style="text-align:left;">
WBP2NL
</td>
<td style="text-align:right;">
24.71140
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
7.00e-07
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
7.00e-07
</td>
<td style="text-align:left;">
rs2301521
</td>
<td style="text-align:right;">
1.0e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
206
</td>
<td style="text-align:left;">
NDUFA6
</td>
<td style="text-align:right;">
24.39774
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
8.00e-07
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
8.00e-07
</td>
<td style="text-align:left;">
rs1801311
</td>
<td style="text-align:right;">
2.0e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
365
</td>
<td style="text-align:left;">
ZBED4
</td>
<td style="text-align:right;">
19.38566
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.07e-05
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
PFS
</td>
<td style="text-align:right;">
1.07e-05
</td>
<td style="text-align:left;">
rs910799
</td>
<td style="text-align:right;">
3.9e-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/dleelab" class="external-link">Donghyung Lee</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
