# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Calculate population weights using RAF
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and af1 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param interval number of non-overlapping SNP sets used in calculating population weights 
#' @return R data frame containing population IDs and weights 
afmix <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL) {
    .Call(`_gauss_afmix`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval)
}

#' Compute LD for measured SNPs from mixed ethnicity cohorts
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param pop_wgt_df R data frame containing population IDs and weights
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R list containing a data frame containing rsid, chr, bp, a1, a2 and af1mix and a correlation matrix 
computeLD <- function(chr, start_bp, end_bp, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_computeLD`, chr, start_bp, end_bp, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Calculate population weights using RAF
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and af1 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param interval number of non-overlapping SNP sets used in calculating population weights 
#' @return R data frame containing population IDs and weights 
cpw2 <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL) {
    .Call(`_gauss_cpw2`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval)
}

#' Direct imputation of summary statistics for unmeasured SNPs from ethnically homogeneous cohorts
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param study_pop study population group
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing rsid, chr, bp, a1, a2, af1ref, z, pval, info, type   
dist <- function(chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_dist`, chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Direct imputation of summary statistics for unmeasured SNPs from mixed ethnicity cohorts
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param pop_wgt_df R data frame containing population IDs and weights
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R data frame containing rsid, chr, bp, a1, a2, af1mix, z, pval, info, type 
distmix <- function(chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_distmix`, chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Joint effect on phenotype of eQTLs/functional SNPs associated with a gene
#' 
#' @param study_pop study population group
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2 and z
#' @param annotation file name of the SNP annotation data set 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing geneid, chisq, df, jepeg_pval, num_snp, top_categ, top_categ_pval, top_snp, and top_snp_pval   
jepeg <- function(study_pop, input_file, annotation_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_jepeg`, study_pop, input_file, annotation_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' JEPEGMIX: gene-level joint analysis of functional SNPs in cosmopolitan cohorts
#' 
#' @param pop_wgt_df R data frame containing population IDs and weights
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2 and z
#' @param annotation file name of the SNP annotation data set 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing geneid, chisq, df, jepeg_pval, num_snp, top_categ, top_categ_pval, top_snp, and top_snp_pval   
jepegmix <- function(pop_wgt_df, input_file, annotation_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_jepegmix`, pop_wgt_df, input_file, annotation_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Prepare datasets for QCAT analysis
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param study_pop study population group
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing rsid, chr, bp, a1, a2, af1ref, z, qcat_chisq, qcat_pval, type
prep_qcat <- function(chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_prep_qcat`, chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Prepare datasets for QCATMIX analysis
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param pop_wgt_df R data frame containing population IDs and weights
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing rsid, chr, bp, a1, a2, af1mix, z, qcat_chisq, qcat_pval, type
prep_qcatmix <- function(chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_prep_qcatmix`, chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Testing causality of variants from ethnically homogeneous cohorts
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param study_pop study population group
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing rsid, chr, bp, a1, a2, af1ref, z, qcat_chisq, qcat_pval, type
qcat <- function(chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_qcat`, chr, start_bp, end_bp, wing_size, study_pop, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Testing causality of variants from mixed ethnicity cohorts
#' 
#' @param chr chromosome number
#' @param start_bp start base pair position of prediction window
#' @param end_bp end base pair position of prediction window
#' @param wing_size the size of the area flanking the left and right of the prediction window
#' @param pop_wgt_df R data frame containing population IDs and weights
#' @param input_file file name of GWAS summary statistics data containing rsid, chr, bp, a1, a2, af1, and z
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
#' @param af1_cutoff cutoff of reference allele, a1, frequency
#' @return R dataframe containing rsid, chr, bp, a1, a2, af1mix, z, qcat_chisq, qcat_pval, type
qcatmix <- function(chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff = NULL) {
    .Call(`_gauss_qcatmix`, chr, start_bp, end_bp, wing_size, pop_wgt_df, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, af1_cutoff)
}

#' Calculate population weights using association Z-scores
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and z 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
NULL

#' Calculate population weights using association Z-scores
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and z 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
NULL

#' Calculate population weights using association Z-scores
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and z 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
NULL

#' Calculate population weights using association Z-scores
#' 
#' @param input_file file name of input data containing rsid, chr, bp, a1, a2, and z 
#' @param reference_index_file file name of reference panel index data
#' @param reference_data_file  file name of reference panel data
#' @param reference_pop_desc_file file name of reference panel population description data
NULL

#' @return R data frame containing population IDs and weights 
zmix4 <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL, offset = NULL) {
    .Call(`_gauss_zmix4`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval, offset)
}

#' @return R data frame containing population IDs and weights 
zmix3 <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL, steps = NULL) {
    .Call(`_gauss_zmix3`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval, steps)
}

#' @return R data frame containing population IDs and weights 
zmix2 <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL, offset = NULL) {
    .Call(`_gauss_zmix2`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval, offset)
}

#' @return R data frame containing population IDs and weights 
zmix <- function(input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval = NULL) {
    .Call(`_gauss_zmix`, input_file, reference_index_file, reference_data_file, reference_pop_desc_file, interval)
}

zmixRcpp <- function(mat, reference_pop_desc_file) {
    .Call(`_gauss_zmixRcpp`, mat, reference_pop_desc_file)
}

